### ISSUE TITLE
Linking getStateFromPath returning undefined when providing config

### ISSUE THREAD

**Issue by NTonani** [react-navigation/react-navigation#9081]
**Current Behavior**
I'm troubleshooting why `initialRouteName` isn't being respected within Linking, so I've overridden `getStateFromPath` to output state object as I suspected the `Events` route was missing:
```
getStateFromPath: (path, options) => {
    logger.info(getStateFromPath(path, options));
    return getStateFromPath(path, options);
}
```

This returns `undefined` when I provide the following config: 
```
        config: {
            screens: {
                Home: {
                    path: 'Home',
                    initialRouteName: 'Events',
                    screens: {
                        Events: 'Events',
                        EventDetails: 'EventDetails',
                    },
                },
            },
        },
```

It returns a state object w/o `Events` route if I don't provide config (of course). 

This same (path, config) pair provides correct state obj in [Playground](https://reactnavigation.org/docs/configuring-links#playground)
```
{
  "routes": [
    {
      "name": "Home",
      "state": {
        "index": 1,
        "routes": [
          {
            "name": "Events"
          },
          {
            "name": "EventDetails"
          }
        ]
      }
    }
  ]
}
```

I quickly stepped through `getStateFromPath` and while the `initialRouteName` is added to routes, everything is lost somehwere [here](https://github.com/react-navigation/react-navigation/blob/%40react-navigation/native%405.8.9/packages/core/src/getStateFromPath.tsx#L229), finally returning `undefined` [here](https://github.com/react-navigation/react-navigation/blob/%40react-navigation/native%405.8.9/packages/core/src/getStateFromPath.tsx#L283). Haven't spent enough time to fully understand what's going on there. Figured I'd raise this instead. 

**Expected Behavior**

Same state output as Playground above ^

**Your Environment**

| software                       | version |
| ------------------------------ | ------- |
| iOS or Android                 | iOS 13.7
| @react-navigation/native       | 5.8.9
| react-native                   | expo 39.0.3
| expo                           | expo 39.0.3
| node                           | 14.7
| npm or yarn                    | 6.14.7

**Comment by kulyk** [comment]
Had the same problem, but I found a tricky way to solve it and it's probably specific to Expo

## TLDR

Try changing your linking prefixes list to:

```javascript
import Constants, {AppOwnership} from 'expo-constants';
import * as Linking from 'expo-linking';

const isStandalone = Constants.appOwnership === AppOwnership.Standalone

<NavigationContainer
   linking={{
        prefixes: [
            Linking.makeURL(isStandalone ? '' : '/'),
            // your associated domains, etc.
        ],
   }}
>
...
</NavigationContainer >
```

## Detailed Description

I hope it would be helpful for react-navigation mainteners

When configuring deep linking in the Expo project, you have different application prefixes in debug (Expo Client) and in _release_ (standalone app)

* for Expo Client apps it's something like `exp://127.0.0.1:19000`
* for standalone app it's the `scheme` from `app.json` (let it be `myapp://` for example)

The [react-navigation docs](https://reactnavigation.org/docs/navigation-container#uri-prefix) suggests to add `Linking.makeUrl('/')` to `NavigationContainer` `linking` config.

When I was working on deep links in my app, I had a button calling `Linking.openUrl(Linking.makeURL('some-url'))` for testing purposes.

So I noticed when I have `makeURL('/')` in linking prefixes, it works in Expo Client. But in a standalone app, nothing happens when I tap the button (because it generates a link like `myapp:///some-url` with three slashes).

I replaced `makeURL('/')` with `makeURL('')` in the linking prefixes and did the `expo publish` to my own release channel. The standalone app started handling deep links (now `makeURL('some-url')` return `myapp://some-url` with two slashes). But the debug version in my Expo Client appeared to be broken

I noticed that `makeURL('some-url')` I use for debugging returns a link like `exp://127.0.0.1:19000/--/some-url` (with `/--/`). I noticed react-navigation's `getStateFromPath` returned `undefined` in the `useLinking.native.ts`.

I've tried to overwrite the `getStateFromPath` just to see what happens:

```javascript

import {getStateFromPath} from '@react-navigation/native';

linking: {
    prefixes: [Linking.makeURL('')],
    getStateFromPath: (path, opts) => {
        console.log(path)
        console.log(opts)
        console.log(getStateFromPath(path, opts))

        // Just returning the same thing react-navigation would return
        return getStateFromPath(path, opts)
    },
},
```

So when `Linking.makeURL('')` is in prefixes, and we try to open `Linking.makeURL('some-url')`, the `path` `getStateFromPath` recieves is `/--/some-url` and it **does not have a mach in linking config**.

When react-navigation receives a link, it goes through `prefixes` in order to cut the **prefix**. And as I use `Linking.makeURL('')` here, the `/--/`, **specific to Expo Client only**, does not get cut.

So, to have deep linking working in both Expo Client and standalone apps, I had to pass different prefixes depending on the `Constants.appOwnership`, so:

* I use `Linking.makeURL('/')` for Expo Client (`exp://127.0.0.1:19000/--/`, but would be `myapp:///` in standalone)
* I use `Linking.makeURL('')` in standaline app (`myapp://`, but would be `exp://127.0.0.1:19000` in Expo Client)

**Comment by satya164** [comment]
Thanks @kulyk for the detailed analysis. It's very helpful.

**Comment by phatmann** [comment]
The [documentation](https://reactnavigation.org/docs/deep-linking#uri-prefix) should be updated. It mistakenly says that createURL has different behavior than makeUrl, and it implies that the use case above is handled. But this is not the case. I will file a new issue.
